org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/GradlePreferenceChangeListener.java

```diff
@@ -0,0 +1,136 @@
+		ProjectsManager projectsManager = JavaLanguageServerPlugin.getProjectsManager();
+		if (projectsManager != null) {
+			if (hasAllowedChecksumsChanged(oldPreferences, newPreferences)) {
+				for (IProject project : ProjectUtils.getGradleProjects()) {
+					updateProject(projectsManager, project);
+				}
+			}
+		}
+	}
+
+	private void updateProject(ProjectsManager projectsManager, IProject project) {
+		String projectDir = project.getLocation().toFile().getAbsolutePath();
+		Path projectPath = Paths.get(projectDir);
+		PreferenceManager preferencesManager = getPreferenceManager();
+		if (preferencesManager != null && preferencesManager.getPreferences().isGradleWrapperEnabled() && Files.exists(projectPath.resolve("gradlew"))) {
+			ProjectConfiguration configuration = CorePlugin.configurationManager().loadProjectConfiguration(project);
+			GradleDistribution distribution = configuration.getBuildConfiguration().getGradleDistribution();
+			if (!(distribution instanceof WrapperGradleDistribution)) {
+				projectsManager.updateProject(project, true);
+			} else {
+				try {
+					ValidationResult result = new WrapperValidator().checkWrapper(projectDir);
+					if (!result.isValid()) {
+						projectsManager.updateProject(project, true);
+					}
+				} catch (CoreException e) {
+					JavaLanguageServerPlugin.logException(e.getMessage(), e);
+				}
+			}
+		}
+	}
+
+	private PreferenceManager getPreferenceManager() {
+		PreferenceManager preferencesManager = JavaLanguageServerPlugin.getPreferencesManager();
+		return preferencesManager;
+	}
+
+	private boolean hasAllowedChecksumsChanged(Preferences oldPreferences, Preferences newPreferences) {
+		if (!Objects.equals(oldPreferences.getGradleWrapperList(), newPreferences.getGradleWrapperList())) {
+			putSha256(newPreferences.getGradleWrapperList());
+			return true;
+		}
+		return false;
+	}
+
+	private void putSha256(List<?> gradleWrapperList) {
+		List<String> sha256Allowed = new ArrayList<>();
+		List<String> sha256Disallowed = new ArrayList<>();
+		for (Object object : gradleWrapperList) {
+			if (object instanceof Map) {
+				Map<?, ?> map = (Map<?, ?>) object;
+				final ChecksumWrapper sha256 = this.new ChecksumWrapper();
+				sha256.allowed = true;
+				map.forEach((k, v) -> {
+					if (k instanceof String) {
+						switch ((String) k) {
+							case "sha256":
+								if (v instanceof String) {
+									sha256.checksum = (String) v;
+								}
+								break;
+							case "allowed":
+								if (v instanceof Boolean) {
+									sha256.allowed = (Boolean) v;
+								}
+								break;
+							default:
+								break;
+						}
+					}
+				});
+				if (sha256.checksum != null) {
+					if (sha256.allowed) {
+						sha256Allowed.add(sha256.checksum);
+					} else {
+						sha256Disallowed.add(sha256.checksum);
+					}
+				}
+			}
+		}
+		WrapperValidator.clear();
+		getPreferenceManager().getPreferences().setSha256Allowed(sha256Allowed);
```

Comment:
That seems wrong. You're updating the preferences in the preference listener!
In here only the WrapperValidator should be updated, from the informations available from the Preference object.

The GradleWrapperList should have been transformed into (dis)allowed checksums directly when building the Preference object.

Reply:
Fixed.